### PR TITLE
refactor: bump aws sdk, use rustls in s3 client

### LINKED ISSUE #9503: discussion: may change `native-tls` to `rustls` to let GCS use http/2
GCS cannot use HTTP/2 under `native-tls` (because there is no ALPN support), may affect object storage performance.
We can rethink this issue after we decided that better gcs performance was required in the future.

### PR DESCRIPTION
**Pull request by MrCroxx**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Bump aws sdk version to latest. Replace `native-tls` with `rustls` for s3 client because aws sdk doesn't support it anymore.

close #9503 #11842 #11852 

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/object_store/src/object/s3.rs [hunk 1]
```diff
@@ -537,41 +540,37 @@ impl S3ObjectStore {
...
+            .enable_all_versions()
+            .wrap_connector(http);
 
-        aws_smithy_client::hyper_ext::Adapter::builder()
-            .hyper_builder(hyper::client::Builder::default())
-            .connector_settings(ConnectorSettings::builder().build())
-            .build(monitor_connector(native_tls, "S3"))
+        let conn = monitor_connector(conn, "S3");
+
+        HyperClientBuilder::new().build(conn)
```
Comment (MrCroxx) [inline src/object_store/src/object/s3.rs:573]: @wenym1 @zwang28 PTAL

File: Cargo.lock [hunk 2]
```diff
@@ -11506,3 +11622,8 @@ dependencies = [
  "libc",
  "pkg-config",
 ]
+
+[[patch.unused]]
+name = "getrandom"
+version = "0.2.9"
+source = "git+https://github.com/madsim-rs/getrandom.git?rev=8daf97e#8daf97e4142635fe28543b2db9022f5e2544bb5c"
```
Comment (xxchan) [inline Cargo.lock:11629]: 👀

Comment (MrCroxx) [reply Cargo.lock:11629]: Yep, madsim also needs to update the dependency. I'll change to madsim source when @wangrunji0408 publishes the new version.

File: src/object_store/Cargo.toml [hunk 3]
```diff
@@ -12,21 +12,24 @@ repository = { workspace = true }
 async-trait = "0.1"
 await-tree = { workspace = true }
 aws-config = { workspace = true }
-aws-sdk-s3 = { version = "0.2", package = "madsim-aws-sdk-s3" }
-aws-smithy-client = { workspace = true }
+aws-sdk-s3 = { version = "0.2", package = "madsim-aws-sdk-s3", git = "https://github.com/mrcroxx/madsim.git", rev = "8efd1d5" }
```
[the remark this answers is not in the document]
Comment (MrCroxx) [reply src/object_store/Cargo.toml:15]: It's temporary. @wanglei4687 will publish a new madsim version then I will fix it.

[the remark this answers is not in the document]
Comment (wanglei4687) [reply src/object_store/Cargo.toml:15]: > It's temporary. @wanglei4687 will publish a new madsim version then I will fix it.

?-? --- Why me?

File: Cargo.lock (hunk 2 above)
Comment (wangrunji0408) [reply Cargo.lock:11629]: published as madsim-aws-sdk-s3 v0.3.0

Comment (wangrunji0408) [reply Cargo.lock:11629]: Btw, getrandom has updated to v0.2.11 some days ago. Although there's nothing important according to its [changelog](https://github.com/rust-random/getrandom/blob/master/CHANGELOG.md). If we are going to bump its version, remember to bump the patch version as well:

```toml
getrandom = { git = "https://github.com/madsim-rs/getrandom.git", rev = "e79a7ae" }
```

File: src/object_store/Cargo.toml (hunk 3 above)
[the remark this answers is not in the document]
Comment (MrCroxx) [reply src/object_store/Cargo.toml:15]: Sorry 🤣 wrong person. @wangrunji0408

[the remark this answers is not in the document]
Comment (MrCroxx) [reply src/object_store/Cargo.toml:15]: Strange. Why wanglei appears in my recommended mention list under this PR. 🤣

[the remark this answers is not in the document]
Comment (MrCroxx) [reply src/object_store/Cargo.toml:15]: Oh, he is a contributor. Thanks for contributing. 🤣

File: Cargo.toml [hunk 4]
```diff
@@ -67,37 +67,30 @@ repository = "https://github.com/risingwavelabs/risingwave"
...
-aws-sdk-ec2 = { version = "0.28", default-features = false, features = [
+aws-sdk-ec2 = { version = "0.35", default-features = false, features = [
   "rt-tokio",
-  "native-tls",
 ] }
-aws-sdk-sqs = { version = "0.28", default-features = false, features = [
+aws-sdk-sqs = { version = "0.35", default-features = false, features = [
   "rt-tokio",
-  "native-tls",
 ] }
```
[the remark this answers is not in the document]
Comment (MrCroxx) [reply Cargo.toml:87]: Fixed. Is there any way to test if rustls is enabled and https is working?

File: src/object_store/src/object/s3.rs [hunk 5]
```diff
@@ -293,7 +294,9 @@ impl StreamingUploader for S3StreamingUploader {
 
 fn get_upload_body(data: Vec<Bytes>) -> ByteStream {
     SdkBody::retryable(move || {
-        Body::wrap_stream(stream::iter(data.clone().into_iter().map(ObjectResult::Ok))).into()
+        SdkBody::from_body_0_4(Body::wrap_stream(stream::iter(
```
Comment (hzxa21) [inline src/object_store/src/object/s3.rs:297]: where does `from_body_0_4` come from? Any documentation?

**Review by hzxa21 — APPROVED** [review]
Rest LGTM

File: Cargo.toml (hunk 4 above)
[the remark this answers is not in the document]
Comment (hzxa21) [reply Cargo.toml]: If you try to connect to S3 via ec2 in the same region without setting the endpoint explicitly, I think https is used by default. You can also set `RW_S3_ENDPOINT` explicitly with `https://....`.

File: src/object_store/src/object/s3.rs (hunk 5 above)
Comment (MrCroxx) [reply src/object_store/src/object/s3.rs]: Fixed. Now it uses `hyper-0-14-x`, which matches our deps.

**Comment by TennyZhuang** [comment]
The PR bumped `getrandom`, then we should also update the madsim patch.

https://github.com/risingwavelabs/risingwave/blob/2df34ee858dae99227f4957585671cc09c736eb5/Cargo.toml#L244

cc @wangrunji0408

File: Cargo.lock (hunk 2 above)
Comment (TennyZhuang) [reply Cargo.lock:11684]: > If we are going to bump its version, remember to bump the patch version as well:

It was eventually forgotten.

Comment (wangrunji0408) [reply Cargo.lock:11684]: 🤣

**Pull request merged**